Source/WebCore/fileapi/ThreadableBlobRegistry.cpp

```diff
@@ -177,35 +178,40 @@ unsigned long long ThreadableBlobRegistry::blobSize(const URL& url)
 
 void ThreadableBlobRegistry::unregisterBlobURL(const URL& url)
 {
-    unregisterBlobURLOriginIfNecessary(url);
-
     ensureOnMainThread([url = url.isolatedCopy()] {
+        unregisterBlobURLOriginIfNecessaryOnMainThread(url);
         blobRegistry().unregisterBlobURL(url);
     });
 }
 
 void ThreadableBlobRegistry::registerBlobURLHandle(const URL& url)
 {
-    if (isBlobURLContainsNullOrigin(url))
-        blobURLReferencesMap()->add(url.stringWithoutFragmentIdentifier());
-
     ensureOnMainThread([url = url.isolatedCopy()] {
+        if (isBlobURLContainsNullOrigin(url))
+            blobURLReferencesMap().add(url.stringWithoutFragmentIdentifier());
+
         blobRegistry().registerBlobURLHandle(url);
     });
 }
 
 void ThreadableBlobRegistry::unregisterBlobURLHandle(const URL& url)
 {
-    unregisterBlobURLOriginIfNecessary(url);
-
     ensureOnMainThread([url = url.isolatedCopy()] {
+        unregisterBlobURLOriginIfNecessaryOnMainThread(url);
         blobRegistry().unregisterBlobURLHandle(url);
     });
 }
 
 RefPtr<SecurityOrigin> ThreadableBlobRegistry::getCachedOrigin(const URL& url)
 {
-    if (auto cachedOrigin = originMap()->get<StringViewHashTranslator>(url.viewWithoutFragmentIdentifier()))
+    RefPtr<SecurityOrigin> cachedOrigin;
+
+    bool wasOnMainThread = isMainThread();
+    callOnMainThreadAndWait([url = url.isolatedCopy(), wasOnMainThread, &cachedOrigin] {
+        if (auto* origin = originMap().get<StringViewHashTranslator>(url.viewWithoutFragmentIdentifier()))
+            cachedOrigin = wasOnMainThread ? Ref { *origin } : origin->isolatedCopy();
```

Comment:
Many SecurityOrigin-related function (such as canRequest() & isSameOriginAs()) have fast paths that do pointer comparison. It is therefore useful to avoid the isolatedCopy() when possible. I didn't see other places in this file where we were isolated copying the origin, except inside registerBlobURL(), when I made sure to only do it when strictly needed as well.